File: hyprexpo-switcher/src/workspace_utils.cpp
#include "workspace_utils.hpp"

#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>

#include <algorithm>
#include <limits>
#include <unordered_map>

namespace {
    struct SWorkspaceClientInfo {
        QString title;
        QString appClass;
        int     windowCount      = 0;
        int     focusHistoryRank = std::numeric_limits<int>::max();
    };

    int mruRankForWorkspace(const std::vector<int>& mruWorkspaceIDs, int workspaceID) {
        for (size_t i = 0; i < mruWorkspaceIDs.size(); ++i) {
            if (mruWorkspaceIDs[i] == workspaceID)
                return static_cast<int>(i);
        }

        return std::numeric_limits<int>::max();
    }
}

QVector<SWorkspaceDescriptor> buildWorkspaceDescriptors(const QByteArray& monitorsJSON, const QByteArray& workspacesJSON, const QByteArray& clientsJSON) {
    const auto monitorDoc   = QJsonDocument::fromJson(monitorsJSON);
    const auto workspaceDoc = QJsonDocument::fromJson(workspacesJSON);
    const auto clientsDoc   = QJsonDocument::fromJson(clientsJSON);

    int activeWorkspaceID = -1;
    for (const auto monitorValue : monitorDoc.array()) {
        const auto monitorObject = monitorValue.toObject();
        if (!monitorObject.value(QStringLiteral("focused")).toBool())
            continue;

        activeWorkspaceID = monitorObject.value(QStringLiteral("activeWorkspace")).toObject().value(QStringLiteral("id")).toInt(-1);
        break;
    }

    std::unordered_map<int, SWorkspaceClientInfo> clientInfoByWorkspace;
    for (const auto clientValue : clientsDoc.array()) {
        const auto clientObject = clientValue.toObject();
        if (!clientObject.value(QStringLiteral("mapped")).toBool())
            continue;

        const auto workspaceID = clientObject.value(QStringLiteral("workspace")).toObject().value(QStringLiteral("id")).toInt(0);
        if (workspaceID <= 0)
            continue;

        auto& workspaceInfo = clientInfoByWorkspace[workspaceID];
        workspaceInfo.windowCount++;

        const auto focusHistoryRank = clientObject.value(QStringLiteral("focusHistoryID")).toInt(std::numeric_limits<int>::max());
        if (focusHistoryRank >= workspaceInfo.focusHistoryRank)
            continue;

        workspaceInfo.focusHistoryRank = focusHistoryRank;
        workspaceInfo.title            = clientObject.value(QStringLiteral("title")).toString();
        workspaceInfo.appClass         = clientObject.value(QStringLiteral("class")).toString();
    }

    QVector<SWorkspaceDescriptor> workspaces;
    for (const auto workspaceValue : workspaceDoc.array()) {
        const auto workspaceObject = workspaceValue.toObject();
        const auto workspaceID     = workspaceObject.value(QStringLiteral("id")).toInt(0);
        if (workspaceID <= 0)
            continue;

        const auto windowCount = workspaceObject.value(QStringLiteral("windows")).toInt(0);
        if (windowCount <= 0)
            continue;

        const auto name              = workspaceObject.value(QStringLiteral("name")).toString(QString::number(workspaceID));
        const auto fallbackTitle     = workspaceObject.value(QStringLiteral("lastwindowtitle")).toString();
        const auto clientInfoIt      = clientInfoByWorkspace.find(workspaceID);
        const auto clientInfoPresent = clientInfoIt != clientInfoByWorkspace.end();
        const auto appClass          = clientInfoPresent ? clientInfoIt->second.appClass : QString{};
        const auto subtitle          = clientInfoPresent && !clientInfoIt->second.title.isEmpty()    ? clientInfoIt->second.title :
                                       !appClass.isEmpty()                                            ? appClass :
                                       !fallbackTitle.isEmpty()                                       ? fallbackTitle :
                                                                                                       QStringLiteral("No recent window");
        const auto focusHistoryRank  = clientInfoPresent ? clientInfoIt->second.focusHistoryRank : std::numeric_limits<int>::max();
        const auto derivedCount      = clientInfoPresent ? clientInfoIt->second.windowCount : windowCount;

        workspaces.push_back(SWorkspaceDescriptor{
            .id               = workspaceID,
            .name             = name,
            .subtitle         = subtitle,
            .appClass         = appClass,
            .windowCount      = derivedCount,
            .focusHistoryRank = focusHistoryRank,
            .active           = workspaceID == activeWorkspaceID,
        });
    }

    std::sort(workspaces.begin(), workspaces.end(), [](const auto& left, const auto& right) {
        if (left.active != right.active)
            return left.active;

        if (left.focusHistoryRank != right.focusHistoryRank)
            return left.focusHistoryRank < right.focusHistoryRank;

        return left.id < right.id;
    });

    return workspaces;
}

void sortWorkspacesForSwitcher(QVector<SWorkspaceDescriptor>& workspaces, const std::vector<int>& mruWorkspaceIDs) {
    std::sort(workspaces.begin(), workspaces.end(), [&](const auto& left, const auto& right) {
        if (left.active != right.active)
            return left.active;

        const auto leftMRURank  = mruRankForWorkspace(mruWorkspaceIDs, left.id);
        const auto rightMRURank = mruRankForWorkspace(mruWorkspaceIDs, right.id);

        if (leftMRURank != rightMRURank)
            return leftMRURank < rightMRURank;

        if (left.focusHistoryRank != right.focusHistoryRank)
            return left.focusHistoryRank < right.focusHistoryRank;

        return left.id < right.id;
    });
}

int initialSelectionIndex(const QVector<SWorkspaceDescriptor>& workspaces, bool reverse) {
    if (workspaces.isEmpty())
        return -1;

    if (workspaces.size() == 1)
        return 0;

    if (reverse)
        return workspaces.size() - 1;

    for (int i = 0; i < workspaces.size(); ++i) {
        if (!workspaces[i].active)
            return i;
    }

    return 0;
}

File: hyprnav/src/workspace_utils.rs
use serde::Deserialize;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct WorkspaceDescriptor {
    pub id: i32,
    pub name: String,
    pub subtitle: String,
    pub app_class: String,
    pub window_count: i32,
    pub focus_history_rank: i32,
    pub active: bool,
}

#[derive(Debug, Deserialize)]
struct MonitorInfo {
    #[serde(default)]
    focused: bool,
    #[serde(default, rename = "activeWorkspace")]
    active_workspace: WorkspaceRef,
}

#[derive(Debug, Default, Deserialize)]
struct WorkspaceRef {
    #[serde(default)]
    id: i32,
}

#[derive(Debug, Deserialize)]
struct WorkspaceInfo {
    #[serde(default)]
    id: i32,
    #[serde(default)]
    name: String,
    #[serde(default)]
    windows: i32,
    #[serde(default, rename = "lastwindowtitle")]
    last_window_title: String,
}

#[derive(Debug, Deserialize)]
struct ClientInfo {
    #[serde(default)]
    mapped: bool,
    #[serde(default)]
    workspace: WorkspaceRef,
    #[serde(default)]
    title: String,
    #[serde(default, rename = "class")]
    app_class: String,
    #[serde(default = "max_i32", rename = "focusHistoryID")]
    focus_history_id: i32,
}

#[derive(Clone, Debug, Default)]
struct WorkspaceClientInfo {
    title: String,
    app_class: String,
    window_count: i32,
    focus_history_rank: i32,
}

pub fn build_workspace_descriptors(
    monitors_json: &[u8],
    workspaces_json: &[u8],
    clients_json: &[u8],
) -> Vec<WorkspaceDescriptor> {
    let monitors = serde_json::from_slice::<Vec<MonitorInfo>>(monitors_json).unwrap_or_default();
    let workspaces =
        serde_json::from_slice::<Vec<WorkspaceInfo>>(workspaces_json).unwrap_or_default();
    let clients = serde_json::from_slice::<Vec<ClientInfo>>(clients_json).unwrap_or_default();

    let active_workspace_id = monitors
        .iter()
        .find(|monitor| monitor.focused)
        .map(|monitor| monitor.active_workspace.id)
        .unwrap_or(-1);

    let mut client_info_by_workspace = std::collections::HashMap::<i32, WorkspaceClientInfo>::new();
    for client in clients
        .into_iter()
        .filter(|client| client.mapped && client.workspace.id > 0)
    {
        let entry = client_info_by_workspace
            .entry(client.workspace.id)
            .or_insert_with(|| WorkspaceClientInfo {
                focus_history_rank: i32::MAX,
                ..WorkspaceClientInfo::default()
            });

        entry.window_count += 1;
        if client.focus_history_id < entry.focus_history_rank {
            entry.focus_history_rank = client.focus_history_id;
            entry.title = client.title;
            entry.app_class = client.app_class;
        }
    }

    let mut descriptors = workspaces
        .into_iter()
        .filter(|workspace| workspace.id > 0 && workspace.windows > 0)
        .map(|workspace| {
            let client_info = client_info_by_workspace.get(&workspace.id);
            let app_class = client_info
                .map(|item| item.app_class.clone())
                .unwrap_or_default();
            let subtitle = client_info
                .filter(|item| !item.title.is_empty())
                .map(|item| item.title.clone())
                .or_else(|| (!app_class.is_empty()).then_some(app_class.clone()))
                .or_else(|| {
                    (!workspace.last_window_title.is_empty())
                        .then_some(workspace.last_window_title.clone())
                })
                .unwrap_or_else(|| "No recent window".to_owned());

            WorkspaceDescriptor {
                id: workspace.id,
                name: if workspace.name.is_empty() {
                    workspace.id.to_string()
                } else {
                    workspace.name
                },
                subtitle,
                app_class,
                window_count: client_info
                    .map(|item| item.window_count)
                    .unwrap_or(workspace.windows),
                focus_history_rank: client_info
                    .map(|item| item.focus_history_rank)
                    .unwrap_or(i32::MAX),
                active: workspace.id == active_workspace_id,
            }
        })
        .collect::<Vec<_>>();

    descriptors.sort_by(|left, right| {
        left.active
            .cmp(&right.active)
            .reverse()
            .then_with(|| left.focus_history_rank.cmp(&right.focus_history_rank))
            .then_with(|| left.id.cmp(&right.id))
    });

    descriptors
}

pub fn sort_workspaces_for_switcher(
    workspaces: &mut [WorkspaceDescriptor],
    mru_workspace_ids: &[i32],
) {
    workspaces.sort_by(|left, right| {
        left.active
            .cmp(&right.active)
            .reverse()
            .then_with(|| {
                mru_rank(mru_workspace_ids, left.id).cmp(&mru_rank(mru_workspace_ids, right.id))
            })
            .then_with(|| left.focus_history_rank.cmp(&right.focus_history_rank))
            .then_with(|| left.id.cmp(&right.id))
    });
}

pub fn initial_selection_index(workspaces: &[WorkspaceDescriptor], reverse: bool) -> i32 {
    if workspaces.is_empty() {
        return -1;
    }

    if workspaces.len() == 1 {
        return 0;
    }

    if reverse {
        return workspaces.len() as i32 - 1;
    }

    workspaces
        .iter()
        .position(|workspace| !workspace.active)
        .map(|index| index as i32)
        .unwrap_or(0)
}

fn mru_rank(mru_workspace_ids: &[i32], workspace_id: i32) -> usize {
    mru_workspace_ids
        .iter()
        .position(|candidate| *candidate == workspace_id)
        .unwrap_or(usize::MAX)
}

const fn max_i32() -> i32 {
    i32::MAX
}
